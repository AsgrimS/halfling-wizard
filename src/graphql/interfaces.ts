import {
  Class,
  Subclass,
  Proficiency,
  ClassProficiency_Choices,
  AbilityScoreSkills,
  ClassSpellcasting,
  ClassStarting_Equipment,
} from "./schema";

export interface HeroClassesData {
  classes: Class[];
}

export interface HeroClassData {
  class: Class;
}

export interface SubclassesData {
  subclasses: Subclass[];
}

export interface ProficienciesData {
  proficiencies: Proficiency[];
}

export interface ProficienciesChoicesData {
  proficienciesChoices: ClassProficiency_Choices[];
}
export interface SavingThrowsData {
  saving_throws: AbilityScoreSkills[];
}

export interface ClassSpellcastingData {
  spellcasting: ClassSpellcasting;
}

export interface StartingEquipmentData {
  startingEquipments: ClassStarting_Equipment[];
}
