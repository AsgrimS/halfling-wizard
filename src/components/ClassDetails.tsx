import React, { useEffect } from "react";

import styled from "styled-components";

import { HeroClassData } from "../graphql/interfaces";
import SubclassesList from "./SubclassesList";
import ProficienciesList from "./ProficienciesList";
import SavingThrowsList from "./SavingThrowsList";
import Spellcasting from "./Spellcasting";
import StartingEquipment from "./StartingEquipment";
import Container from "./styled-components/Container";

const ClassDetails: React.FC<HeroClassData> = ({ class: heroClass }) => {
  useEffect(() => {
    document.getElementById("class-details")?.scrollIntoView();
  }, [heroClass.name]);

  return (
    <Details id="class-details">
      <div>
        <span>{heroClass.name}</span>
      </div>
      <div>
        <span>Hit die {heroClass.hit_die}</span>
      </div>
      {heroClass.subclasses && (
        <SubclassesList subclasses={heroClass.subclasses} />
      )}
      {heroClass.saving_throws && (
        <SavingThrowsList saving_throws={heroClass.saving_throws} />
      )}
      {heroClass.spellcasting && (
        <Spellcasting spellcasting={heroClass.spellcasting} />
      )}
      {heroClass.proficiencies && heroClass.proficiency_choices && (
        <ProficienciesList
          proficiencies={heroClass.proficiencies}
          proficienciesChoices={heroClass.proficiency_choices}
        />
      )}
      {heroClass.starting_equipment && (
        <StartingEquipment startingEquipments={heroClass.starting_equipment} />
      )}
      <PushDown />
    </Details>
  );
};

export default ClassDetails;

const Details = styled(Container)`
  display: flex;
  flex-direction: column;
  overflow-y: auto;
  --backgroundColor: var(--blackOliveL);

  & > div:not(:last-of-type) {
    background-color: var(--eerieBlack);
    opacity: 0.9;
    border-radius: 0.5rem;
    margin-bottom: 0.5rem;
    min-width: 20ch;
  }
`;

const PushDown = styled.div`
  margin-bottom: 0;
  :after {
    content: "";
    display: block;
    height: 2.5vh;
    width: 100%;
  }
`;
