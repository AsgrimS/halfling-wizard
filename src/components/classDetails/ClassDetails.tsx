import React from "react";

import styled from "styled-components";

import { HeroClassData } from "../../graphql/interfaces";
import SubclassesList from "../subclassesList/SubclassesList";

const ClassDetails: React.FC<HeroClassData> = ({ class: heroClass }) => {
  return (
    <Details>
      <p>{heroClass.name}</p>
      <p>{heroClass.hit_die}</p>
      {heroClass.subclasses && (
        <SubclassesList subclasses={heroClass.subclasses} />
      )}
    </Details>
  );
};

export default ClassDetails;

const Details = styled.div`
  grid-column-start: 2;
  grid-column-end: 2;
`;
