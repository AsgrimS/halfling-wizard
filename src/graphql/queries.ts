import { gql } from "@apollo/client";

export const HERO_CLASS = gql`
  query GetHeroesClasses($classIndex: String!) {
    class(filter: { name: $classIndex }) {
      hit_die
      index
      name
      proficiencies {
        index
        name
      }
      saving_throws {
        index
        name
      }
      proficiency_choices {
        choose
        from {
          name
          index
        }
      }
      spellcasting {
        level
        spellcasting_ability {
          index
          name
        }
      }
      starting_equipment {
        equipment {
          name
          index
        }
        quantity
      }
      starting_equipment_options {
        choose
        from {
          equipment {
            name
            index
          }
          quantity
        }
      }
      subclasses {
        index
        name
      }
    }
  }
`;

export const HEROES_CLASSES = gql`
  query GetHeroesClasses {
    classes {
      name
      index
    }
  }
`;
