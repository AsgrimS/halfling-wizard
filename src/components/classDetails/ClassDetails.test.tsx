import React from "react";
import { render, screen, cleanup } from "@testing-library/react";

import ClassDetails from "./ClassDetails";
import { Class } from "../../graphql/schema";

jest.mock("../subclassesList/SubclassesList", () => () => (
  <div data-testid="mock-subclasses-list" />
));
jest.mock("../proficienciesList/ProficienciesList", () => () => (
  <div data-testid="mock-proficiencies-list" />
));
jest.mock("../proficienciesChoicesList/ProficienciesChoicesList", () => () => (
  <div data-testid="mock-proficiencies-choices-list" />
));
jest.mock("../savingThrowsList/SavingThrowsList", () => () => (
  <div data-testid="mock-saving-throws" />
));
jest.mock("../spellcasting/Spellcasting", () => () => (
  <div data-testid="mock-spellcasting" />
));
jest.mock("../startingEquipment/StartingEquipment", () => () => (
  <div data-testid="mock-starting-equipment" />
));

describe("ClassDetails component", () => {
  afterEach(() => {
    cleanup();
  });

  it("should render a component that displays details for class", () => {
    const heroClassMockData: Class = {
      name: "Wizard",
      hit_die: 6,
      subclasses: [{ name: "Dark Wizard" }],
      saving_throws: [{ name: "LUCK" }],
      spellcasting: { spellcasting_ability: { name: "YES" } },
      starting_equipment: [{ quantity: 1, equipment: { name: "stick" } }],
      proficiencies: [{ name: "writing mock data" }],
      proficiency_choices: [
        { choose: 1, from: [{ name: "it was not a choice" }] },
      ],
    };
    const scrollIntoViewSpy = jest.fn();
    window.HTMLElement.prototype.scrollIntoView = scrollIntoViewSpy;

    render(<ClassDetails class={heroClassMockData} />);
    const classDetailsElement = screen.getByTestId("class-details");
    const spanElements = classDetailsElement.getElementsByTagName("span");

    expect(classDetailsElement).toBeInTheDocument();
    expect(spanElements.length).toEqual(2);

    expect(spanElements[0]).toContainHTML(
      `<span>${heroClassMockData.name}</span>`
    );
    expect(spanElements[1]).toContainHTML(
      `<span>Hit die ${heroClassMockData.hit_die}</span>`
    );

    const mockedComponents = [
      screen.getByTestId("mock-subclasses-list"),
      screen.getByTestId("mock-proficiencies-list"),
      screen.getByTestId("mock-proficiencies-choices-list"),
      screen.getByTestId("mock-saving-throws"),
      screen.getByTestId("mock-spellcasting"),
      screen.getByTestId("mock-starting-equipment"),
    ];

    mockedComponents.forEach((mockedComponent) => {
      expect(mockedComponent).toBeInTheDocument();
    });

    expect(classDetailsElement).toMatchSnapshot();
  });
});
