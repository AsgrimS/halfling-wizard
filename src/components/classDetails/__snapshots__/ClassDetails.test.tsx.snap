// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClassDetails component should render a component that displays details for class 1`] = `
<section
  class="sc-bdnxRM sc-gtsrHT bgsJOG hGOABx"
  data-testid="class-details"
  id="class-details"
>
  <div>
    <span>
      Wizard
    </span>
  </div>
  <div>
    <span>
      Hit die 
      6
    </span>
  </div>
  <div
    data-testid="mock-subclasses-list"
  />
  <div
    data-testid="mock-saving-throws"
  />
  <div
    data-testid="mock-spellcasting"
  />
  <div
    data-testid="mock-starting-equipment"
  />
  <div
    data-testid="mock-proficiencies-list"
  />
  <div
    data-testid="mock-proficiencies-choices-list"
  />
  <div
    class="sc-dlnjwi cWeluP"
  />
</section>
`;
